support/releaseinfo/registryclient/client.go

```diff
@@ -123,3 +100,159 @@ func allFound(content map[string][]byte) bool {
+		err := func() error {
+			r, err := fromBlobs.Open(ctx, layer.Digest)
+			if err != nil {
+				return fmt.Errorf("unable to access the source layer %s: %v", layer.Digest, err)
+			}
+			defer r.Close()
+			rc, err := dockerarchive.DecompressStream(r)
+			if err != nil {
+				return err
+			}
+			defer rc.Close()
+			tr := tar.NewReader(rc)
+			for {
+				hdr, err := tr.Next()
+				if err != nil {
+					if err == io.EOF {
+						break
+					}
+					return err
+				}
+				if hdr.Typeflag == tar.TypeReg {
+					if hdr.Name != file {
+						continue
+					}
+					found = true
+					if _, err := io.Copy(out, tr); err != nil {
+						return err
+					}
+					return nil
+				}
+			}
+			return nil
+		}()
+		if err != nil {
+			return err
+		}
+		if found {
+			return nil
+		}
+	}
+	return fmt.Errorf("file not found")
+}
+
+func ExtractImageFilesToDir(ctx context.Context, imageRef string, pullSecret []byte, pattern string, outputDir string) error {
+	regex, err := regexp.Compile(pattern)
+	if err != nil {
+		return fmt.Errorf("invalid pattern: %w", err)
+	}
+
+	layers, fromBlobs, err := getMetadata(ctx, imageRef, pullSecret)
+	if err != nil {
+		return err
+	}
+
+	// Iterate over layers in reverse order to find the most recent version of files
+	written := map[string]struct{}{}
+	for i := len(layers) - 1; i >= 0; i-- {
+		layer := layers[i]
+		err := func() error {
+			r, err := fromBlobs.Open(ctx, layer.Digest)
+			if err != nil {
+				return fmt.Errorf("unable to access the source layer %s: %v", layer.Digest, err)
+			}
+			defer r.Close()
+			rc, err := dockerarchive.DecompressStream(r)
+			if err != nil {
+				return err
+			}
+			defer rc.Close()
+			tr := tar.NewReader(rc)
+			for {
+				hdr, err := tr.Next()
+				if err != nil {
+					if err == io.EOF {
+						break
+					}
+					return err
+				}
+				if hdr.Typeflag == tar.TypeReg {
+					// Only copy the file once from the most recent layer
+					if _, exists := written[hdr.Name]; exists {
+						continue
+					}
+					if !regex.MatchString(hdr.Name) {
+						continue
+					}
+					dst := filepath.Join(outputDir, hdr.Name)
+					if err := os.MkdirAll(filepath.Clean(filepath.Dir(dst)), 0755); err != nil {
+						return fmt.Errorf("failed to make dir: %w", err)
+					}
+					dstfd, err := os.Create(dst)
+					if err != nil {
+						return err
+					}
+					if _, err = io.Copy(dstfd, tr); err != nil {
+						dstfd.Close()
+						return err
+					}
+					dstfd.Close()
+					written[hdr.Name] = struct{}{}
+				}
+			}
+			return nil
+		}()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getMetadata(ctx context.Context, imageRef string, pullSecret []byte) ([]distribution.Descriptor, distribution.BlobStore, error) {
+	rt, err := rest.TransportFor(&rest.Config{})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create secure transport: %w", err)
+	}
+	insecureRT, err := rest.TransportFor(&rest.Config{TLSClientConfig: rest.TLSClientConfig{Insecure: true}})
```

Comment:
Luckily immutable Shas are used on the pull so that shouldn’t be a GA blocking problem however we should correct this as this will get flagged in any audit

Reply:
I think that's fair but I think it's a high priority one to close for auditing purposes

Reply:
I suspect we need to take this as a followup, because it's going to need to be configurable so that we can use insecure registries during development, CI scenarios, etc.

Reply:
Cesar corrected me that this has no affect and is simply part of the registry client (it demands that both a insecure client and secure client is provided at initialization. And then a user chooses which is called when parsing the metadata. Our call always uses the secure client and therefore there is no risk here